### PR TITLE
First part of the move of the reporting to the scheduler

### LINKED ISSUE #8189: Make exectutor report deploys to the scheduler
On deploy, the executor reports start and end to the server, we want to move this to the scheduler. 

It should be feasible by:
1. [x] updating executor to not call deploy_start and deploy_done to the server, but do it in the surrounding Deploy task 
2. [ ]  inline the deploy_start and deploy_done methods into the scheduler 
3. [ ] strip them to cover only the case at hand  (when replacing in-db state with in-memory state, be mindful of recovery behavior)

May interact with #8187  and should not be done in parallel

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

closes #8189 

- [x] we no longer set the deploy done field, spin out remove to other ticket #8252 
- [x] spin out ticket to read status of requires from scheduler instead of DB #8253 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/agent/executor.py [hunk 1]
```diff
@@ -471,6 +474,38 @@ async def create_member(self, member_id: EnvBlueprint) -> ExecutorVirtualEnviron
         return process_environment
 
 
+@dataclass
+class DeployResult:
```
Comment (wouterdb) [inline src/inmanta/agent/executor.py:478]: This is the serializable form of the handler context

File: tests/agent_server/deploy/test_scheduler_agent.py [hunk 2]
```diff
@@ -166,6 +193,17 @@ async def join(self, thread_pool_finalizer: list[ThreadPoolExecutor], timeout: f
         pass
 
 
+class DummyStateManager(StateUpdateManager):
+
+    async def send_in_progress(
+        self, action_id: UUID, resource_id: ResourceVersionIdStr
+    ) -> dict[ResourceIdStr, const.ResourceState]:
+        pass
```
Comment (wouterdb) [inline tests/agent_server/deploy/test_scheduler_agent.py:201]: todo: test state update consistency

File: src/inmanta/deploy/tasks.py [hunk 3]
```diff
@@ -109,52 +109,81 @@ async def execute(self, task_manager: "scheduler.TaskManager", agent: str) -> No
 
 class Deploy(Task):
     async def execute(self, task_manager: "scheduler.TaskManager", agent: str) -> None:
```
Comment (wouterdb) [inline src/inmanta/deploy/tasks.py:111]: TODO: make logging double to both DB and logger, as before

File: src/inmanta/deploy/persistence.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+"""
+    Copyright 2024 Inmanta
```
Comment (bartv) [inline src/inmanta/deploy/persistence.py:2]: What happened to all the newlines we normally have?

File: src/inmanta/data/__init__.py [hunk 5]
```diff
@@ -4020,7 +4024,7 @@ class ResourceAction(BaseDocument):
 
     messages: Optional[list[dict[str, object]]] = None
     status: Optional[const.ResourceState] = None
-    changes: Optional[dict[m.ResourceIdStr, dict[str, object]]] = None
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4023]: This was not correct

File: src/inmanta/data/model.py [hunk 6]
```diff
@@ -426,19 +425,21 @@ class ResourceIdDetails(BaseModel):
     resource_id_value: str
 
 
-class OrphanedResource(str, Enum):
+class ReleasedResourceState(StrEnum):
```
Comment (wouterdb) [inline src/inmanta/data/model.py:428]: New mypy didn't like it, me neither

File: tests/deploy/test_scheduler_agent.py [hunk 7]
```diff
@@ -75,13 +80,28 @@ def reset_counters(self) -> None:
...
             else const.ResourceState.deployed
         )
+        return DeployResult(
+            resource_details.rvid,
+            action_id,
+            result,
+            [],
+            {},
+            Change.nochange,
+        )
```
Comment (arnaudsjs) [inline tests/deploy/test_scheduler_agent.py:104]:
```suggestion
        return DeployResult(
            resource_details.rvid,
            action_id,
            status=result,
            messages=[],
            changes={},
            change=Change.nochange,
        )
```

File: tests/deploy/test_scheduler_agent.py [hunk 8]
```diff
@@ -123,14 +143,22 @@ async def stop(self) -> None:
...
+    ) -> DeployResult:
         assert resource_details.rid not in self._deploys
         self._deploys[resource_details.rid] = asyncio.get_running_loop().create_future()
         # wait until the test case sets desired resource state
         result: const.ResourceState = await self._deploys[resource_details.rid]
         del self._deploys[resource_details.rid]
         self.execute_count += 1
-        return result
+
+        return DeployResult(resource_details.rvid, action_id, result, [], {}, Change.nochange)
```
Comment (arnaudsjs) [inline tests/deploy/test_scheduler_agent.py:161]:
```suggestion
        return DeployResult(resource_details.rvid, action_id, status=result, messages=[], changes={}, change=Change.nochange)
```

File: tests/deploy/test_scheduler_agent.py [hunk 9]
```diff
@@ -166,6 +194,48 @@ async def join(self, thread_pool_finalizer: list[ThreadPoolExecutor], timeout: f
         pass
 
 
+state_translation_table = {
```
Comment (arnaudsjs) [inline tests/deploy/test_scheduler_agent.py:197]: This needs a docstring. What does it represent?

File: tests/deploy/test_scheduler_agent.py [hunk 10]
```diff
@@ -166,6 +194,48 @@ async def join(self, thread_pool_finalizer: list[ThreadPoolExecutor], timeout: f
...
+
+    async def send_in_progress(
+        self, action_id: UUID, resource_id: ResourceVersionIdStr
+    ) -> dict[ResourceIdStr, const.ResourceState]:
+        self.state[Id.parse_id(resource_id).resource_str()] = const.ResourceState.deploying
+
+    async def send_deploy_done(self, result: DeployResult) -> None:
+        self.state[Id.parse_id(result.rvid).resource_str()] = result.status
+
+    def check_with_scheduler(self, scheduler: ResourceScheduler) -> None:
```
Comment (arnaudsjs) [inline tests/deploy/test_scheduler_agent.py:223]: This method needs a docstring

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/deploy/persistence.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+    def log_resource_action(
+        self, env: UUID, resource_ids: Sequence[str], log_level: int, ts: datetime.datetime, message: str
+    ) -> None:
+        """Write the given log to the correct resource action logger"""
+        logger = self.get_resource_action_logger(env)
+        if len(resource_ids) == 0:
+            message = "no resources: " + message
+        elif len(resource_ids) > 1:
+            message = "multiple resources: " + message
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/deploy/persistence.py:285]: nope, copy-paste, can't even happen anymore, will fix, thx.

File: tests/deploy/e2e/test_autostarted.py [hunk 12]
```diff
@@ -189,7 +189,6 @@ async def test_auto_deploy_no_splay(server, client, clienthelper: ClientHelper,
     assert result.code == 200
     assert result.result["model"]["released"]
     assert result.result["model"]["total"] == 2
-    assert result.result["model"]["result"] == "failed"
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply tests/deploy/e2e/test_autostarted.py:192]: I stopped updating the state of the model, will be cleaned up in https://github.com/inmanta/inmanta-core/issues/8252

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/agent/in_process_executor.py [hunk 13]
```diff
@@ -296,12 +296,7 @@ async def execute(
             if set_fact_response.code != 200:
                 ctx.error("Failed to send facts to the server %s", set_fact_response.result)
 
-        await self._report_resource_deploy_done(resource_details, ctx)
```
Comment (Hugo-Inmanta) [inline src/inmanta/agent/in_process_executor.py:299]: I think this obsolete method can be removed.

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

**Review by Hugo-Inmanta — APPROVED** [review]

**Pull request closed without merging**